Add adjoint for `map`.

roberthoenig opened the pull request:
See https://github.com/FluxML/Zygote.jl/issues/71 for context.

MikeInnes commented on src/lib/lib.jl:
```diff
@@ -43,6 +43,32 @@ unwrap(x) = x
+# TODO: Replace with more performant version. See https://github.com/JuliaLang/julia/issues/13942.
+unzip(zipped) = collect(zip(zipped...))
+
+function sum_named_tuple(tpl)
```
You should be able to replace this with `reduce(accum, xs)` (`accum` is our generalised `+`).

MikeInnes commented on src/lib/lib.jl:
```diff
@@ -43,6 +43,32 @@ unwrap(x) = x
+
+@adjoint function map(f, args...)
+  ys_and_backs = map((args...) -> Zygote._forward(f, args...), args...)
+  ys, backs = unzip(ys_and_backs)
```
Unzip is pretty slow, in this case you could just do `ys = map(x -> x[1], ys_and_backs` etc.

roberthoenig commented:
Updated: The map adjoint now uses a faster unzip and accum instead of +. Added tests.

Regarding unzip, I wondered whether it'd be beneficial to do the first unzip manually with the map code you suggested. Here are the benchmarks (unzip3 timed out):
![image](https://user-images.githubusercontent.com/7950151/53368425-a4f94300-3940-11e9-8bed-0a8d2093d5a2.png)
Is 3 microseconds difference a lot?

Also, I was wondering why we need a custom adjoint for map in the first place. Wouldn't it be better to have Zygote support the primitives that map is made out of?

roberthoenig commented:
Hmm, not sure what to make of the Travis error for ZYGOTE_TYPED.

MikeInnes commented:
Can you try throwing in a line like `Zygote.usetyped && Zygote.refresh()` before the map tests, to see if the typed version works then?

> Also, I was wondering why we need a custom adjoint for map in the first place. Wouldn't it be better to have Zygote support the primitives that map is made out of?

You are absolutely right, but unfortunately the implementation for `map` will use mutation, which we don't currently support. Even with that it might be a bit faster to take control of Zygote's stacks (the array of backpropagators in this case).

MikeInnes commented on src/lib/array.jl:
```diff
@@ -65,6 +65,22 @@ end
+  end
+end
+@adjoint function map(f, args...)
+  ys_and_backs = map((args...) -> Zygote._forward(f, args...), args...)
```
Almost forgot, you should call `Zygote._forward(__context__, ...)`. The `__context__` is a hidden argument to `@adjoint`s that contains some stuff useful for e.g. differentiating globals.

roberthoenig replied:
Updated!

roberthoenig commented:
Still fails with`usetyped` when I `Zygote.refresh()` before the `map` tests:

```julia
map: Error During Test at /home/robert/.julia/dev/Zygote/test/gradcheck.jl:75
  Test threw exception
  Expression: gradtest((xss...->begin
            sum(map((xs...->begin
                            sqrt(sum(xs .^ 2))
                        end), xss...))
        end), [rand(5) for _ = 1:6]...)
  Compiling Tuple{typeof(ntuple),getfield(Base.Broadcast, Symbol("##17#18")){Base.Broadcast.Broadcasted{Base.Broadcast.Style{Tuple},Nothing,typeof(Base.literal_pow),Tuple{Base.RefValue{typeof(^)},NTuple{6,Float64},Base.RefValue{Val{2}}}}},Val{6}}: type GotoNode has no field val
  Stacktrace:
   [1] getproperty(::Any, ::Symbol) at ./sysimg.jl:18
   [2] forward_stacks!(::Zygote.Adjoint, ::Type) at /home/robert/.julia/dev/Zygote/src/compiler/emit.jl:74
   [3] stacks!(::Zygote.Adjoint, ::Type) at /home/robert/.julia/dev/Zygote/src/compiler/emit.jl:114
   [4] _lookup_grad(::Type) at /home/robert/.julia/dev/Zygote/src/compiler/emit.jl:125
   [5] #s2433#999 at /home/robert/.julia/dev/Zygote/src/compiler/interface2.jl:17 [inlined]
   [6] #s2433#999(::Any, ::Any, ::Any) at ./none:0
   [7] (::Core.GeneratedFunctionStub)(::Any, ::Vararg{Any,N} where N) at ./boot.jl:506
   [8] tuplebroadcast at ./broadcast.jl:922 [inlined]
   [9] (::Zygote.Pullback{Tuple{typeof(Base.Broadcast.tuplebroadcast),NTuple{6,Float64},Base.Broadcast.Broadcasted{Base.Broadcast.Style{Tuple},Nothing,typeof(Base.literal_pow),Tuple{Base.RefValue{typeof(^)},NTuple{6,Float64},Base.RefValue{Val{2}}}}},Any})(::NTuple{6,Float64}) at /home/robert/.julia/dev/Zygote/src/compiler/interface2.jl:0
   [10] copy at ./broadcast.jl:920 [inlined]
   [11] (::Zygote.Pullback{Tuple{typeof(copy),Base.Broadcast.Broadcasted{Base.Broadcast.Style{Tuple},Nothing,typeof(Base.literal_pow),Tuple{Base.RefValue{typeof(^)},NTuple{6,Float64},Base.RefValue{Val{2}}}}},Any})(::NTuple{6,Float64}) at /home/robert/.julia/dev/Zygote/src/compiler/interface2.jl:0
   [12] materialize at ./broadcast.jl:724 [inlined]
   [13] #50 at ./none:0 [inlined]
   [14] (::Zygote.Pullback{Tuple{getfield(Main, Symbol("##50#56")),Float64,Float64,Float64,Float64,Float64,Float64},Tuple{getfield(Zygote, Symbol("##144#back#108")){getfield(Zygote, Symbol("##106#107"))},getfield(Zygote, Symbol("##2386#back#970")){getfield(Zygote, Symbol("##968#969"))},Zygote.Pullback{Tuple{typeof(Base.Broadcast.materialize),Base.Broadcast.Broadcasted{Base.Broadcast.Style{Tuple},Nothing,typeof(Base.literal_pow),Tuple{Base.RefValue{typeof(^)},NTuple{6,Float64},Base.RefValue{Val{2}}}}},Tuple{Zygote.Pullback{Tuple{typeof(Base.Broadcast.instantiate),Base.Broadcast.Broadcasted{Base.Broadcast.Style{Tuple},Nothing,typeof(Base.literal_pow),Tuple{Base.RefValue{typeof(^)},NTuple{6,Float64},Base.RefValue{Val{2}}}}},Tuple{}},Zygote.Pullback{Tuple{typeof(copy),Base.Broadcast.Broadcasted{Base.Broadcast.Style{Tuple},Nothing,typeof(Base.literal_pow),Tuple{Base.RefValue{typeof(^)},NTuple{6,Float64},Base.RefValue{Val{2}}}}},Any}}},Zygote.Pullback{Tuple{typeof(sum),NTuple{6,Float64}},Tuple{getfield(Zygote, Symbol("##240#back#147")){getfield(Zygote, Symbol("##145#146")){getfield(Zygote, Symbol("##1705#back#649")){getfield(Zygote, Symbol("##645#647")){NTuple{6,Float64}}},Tuple{NTuple{6,Nothing}}}}}},getfield(Zygote, Symbol("##1021#back#419")){getfield(Zygote, Symbol("##417#418")){Float64}}}})(::Float64) at /home/robert/.julia/dev/Zygote/src/compiler/interface2.jl:0
   [15] #804 at /home/robert/.julia/dev/Zygote/src/lib/array.jl:77 [inlined]
   [16] #3 at ./generator.jl:36 [inlined]
   [17] iterate at ./generator.jl:47 [inlined]
   [18] collect(::Base.Generator{Base.Iterators.Zip2{Array{Zygote.Pullback{Tuple{getfield(Main, Symbol("##50#56")),Float64,Float64,Float64,Float64,Float64,Float64},Tuple{getfield(Zygote, Symbol("##144#back#108")){getfield(Zygote, Symbol("##106#107"))},getfield(Zygote, Symbol("##2386#back#970")){getfield(Zygote, Symbol("##968#969"))},Zygote.Pullback{Tuple{typeof(Base.Broadcast.materialize),Base.Broadcast.Broadcasted{Base.Broadcast.Style{Tuple},Nothing,typeof(Base.literal_pow),Tuple{Base.RefValue{typeof(^)},NTuple{6,Float64},Base.RefValue{Val{2}}}}},Tuple{Zygote.Pullback{Tuple{typeof(Base.Broadcast.instantiate),Base.Broadcast.Broadcasted{Base.Broadcast.Style{Tuple},Nothing,typeof(Base.literal_pow),Tuple{Base.RefValue{typeof(^)},NTuple{6,Float64},Base.RefValue{Val{2}}}}},Tuple{}},Zygote.Pullback{Tuple{typeof(copy),Base.Broadcast.Broadcasted{Base.Broadcast.Style{Tuple},Nothing,typeof(Base.literal_pow),Tuple{Base.RefValue{typeof(^)},NTuple{6,Float64},Base.RefValue{Val{2}}}}},Any}}},Zygote.Pullback{Tuple{typeof(sum),NTuple{6,Float64}},Tuple{getfield(Zygote, Symbol("##240#back#147")){getfield(Zygote, Symbol("##145#146")){getfield(Zygote, Symbol("##1705#back#649")){getfield(Zygote, Symbol("##645#647")){NTuple{6,Float64}}},Tuple{NTuple{6,Nothing}}}}}},getfield(Zygote, Symbol("##1021#back#419")){getfield(Zygote, Symbol("##417#418")){Float64}}}},1},Zygote.FillArray{Float64,1}},getfield(Base, Symbol("##3#4")){getfield(Zygote, Symbol("##804#807"))}}) at ./array.jl:619
   [19] map at ./abstractarray.jl:2060 [inlined]
   [20] (::getfield(Zygote, Symbol("##803#806")){Array{Zygote.Pullback{Tuple{getfield(Main, Symbol("##50#56")),Float64,Float64,Float64,Float64,Float64,Float64},Tuple{getfield(Zygote, Symbol("##144#back#108")){getfield(Zygote, Symbol("##106#107"))},getfield(Zygote, Symbol("##2386#back#970")){getfield(Zygote, Symbol("##968#969"))},Zygote.Pullback{Tuple{typeof(Base.Broadcast.materialize),Base.Broadcast.Broadcasted{Base.Broadcast.Style{Tuple},Nothing,typeof(Base.literal_pow),Tuple{Base.RefValue{typeof(^)},NTuple{6,Float64},Base.RefValue{Val{2}}}}},Tuple{Zygote.Pullback{Tuple{typeof(Base.Broadcast.instantiate),Base.Broadcast.Broadcasted{Base.Broadcast.Style{Tuple},Nothing,typeof(Base.literal_pow),Tuple{Base.RefValue{typeof(^)},NTuple{6,Float64},Base.RefValue{Val{2}}}}},Tuple{}},Zygote.Pullback{Tuple{typeof(copy),Base.Broadcast.Broadcasted{Base.Broadcast.Style{Tuple},Nothing,typeof(Base.literal_pow),Tuple{Base.RefValue{typeof(^)},NTuple{6,Float64},Base.RefValue{Val{2}}}}},Any}}},Zygote.Pullback{Tuple{typeof(sum),NTuple{6,Float64}},Tuple{getfield(Zygote, Symbol("##240#back#147")){getfield(Zygote, Symbol("##145#146")){getfield(Zygote, Symbol("##1705#back#649")){getfield(Zygote, Symbol("##645#647")){NTuple{6,Float64}}},Tuple{NTuple{6,Nothing}}}}}},getfield(Zygote, Symbol("##1021#back#419")){getfield(Zygote, Symbol("##417#418")){Float64}}}},1}})(::Zygote.FillArray{Float64,1}) at /home/robert/.julia/dev/Zygote/src/lib/array.jl:77
   [21] #2045#back at /home/robert/.julia/dev/Zygote/src/lib/grad.jl:41 [inlined]
   [22] (::getfield(Zygote, Symbol("##145#146")){getfield(Zygote, Symbol("##2045#back#808")){getfield(Zygote, Symbol("##803#806")){Array{Zygote.Pullback{Tuple{getfield(Main, Symbol("##50#56")),Float64,Float64,Float64,Float64,Float64,Float64},Tuple{getfield(Zygote, Symbol("##144#back#108")){getfield(Zygote, Symbol("##106#107"))},getfield(Zygote, Symbol("##2386#back#970")){getfield(Zygote, Symbol("##968#969"))},Zygote.Pullback{Tuple{typeof(Base.Broadcast.materialize),Base.Broadcast.Broadcasted{Base.Broadcast.Style{Tuple},Nothing,typeof(Base.literal_pow),Tuple{Base.RefValue{typeof(^)},NTuple{6,Float64},Base.RefValue{Val{2}}}}},Tuple{Zygote.Pullback{Tuple{typeof(Base.Broadcast.instantiate),Base.Broadcast.Broadcasted{Base.Broadcast.Style{Tuple},Nothing,typeof(Base.literal_pow),Tuple{Base.RefValue{typeof(^)},NTuple{6,Float64},Base.RefValue{Val{2}}}}},Tuple{}},Zygote.Pullback{Tuple{typeof(copy),Base.Broadcast.Broadcasted{Base.Broadcast.Style{Tuple},Nothing,typeof(Base.literal_pow),Tuple{Base.RefValue{typeof(^)},NTuple{6,Float64},Base.RefValue{Val{2}}}}},Any}}},Zygote.Pullback{Tuple{typeof(sum),NTuple{6,Float64}},Tuple{getfield(Zygote, Symbol("##240#back#147")){getfield(Zygote, Symbol("##145#146")){getfield(Zygote, Symbol("##1705#back#649")){getfield(Zygote, Symbol("##645#647")){NTuple{6,Float64}}},Tuple{NTuple{6,Nothing}}}}}},getfield(Zygote, Symbol("##1021#back#419")){getfield(Zygote, Symbol("##417#418")){Float64}}}},1}}},Tuple{Tuple{Nothing},NTuple{6,Nothing}}})(::Zygote.FillArray{Float64,1}) at /home/robert/.julia/dev/Zygote/src/lib/lib.jl:81
   [23] (::getfield(Zygote, Symbol("##240#back#147")){getfield(Zygote, Symbol("##145#146")){getfield(Zygote, Symbol("##2045#back#808")){getfield(Zygote, Symbol("##803#806")){Array{Zygote.Pullback{Tuple{getfield(Main, Symbol("##50#56")),Float64,Float64,Float64,Float64,Float64,Float64},Tuple{getfield(Zygote, Symbol("##144#back#108")){getfield(Zygote, Symbol("##106#107"))},getfield(Zygote, Symbol("##2386#back#970")){getfield(Zygote, Symbol("##968#969"))},Zygote.Pullback{Tuple{typeof(Base.Broadcast.materialize),Base.Broadcast.Broadcasted{Base.Broadcast.Style{Tuple},Nothing,typeof(Base.literal_pow),Tuple{Base.RefValue{typeof(^)},NTuple{6,Float64},Base.RefValue{Val{2}}}}},Tuple{Zygote.Pullback{Tuple{typeof(Base.Broadcast.instantiate),Base.Broadcast.Broadcasted{Base.Broadcast.Style{Tuple},Nothing,typeof(Base.literal_pow),Tuple{Base.RefValue{typeof(^)},NTuple{6,Float64},Base.RefValue{Val{2}}}}},Tuple{}},Zygote.Pullback{Tuple{typeof(copy),Base.Broadcast.Broadcasted{Base.Broadcast.Style{Tuple},Nothing,typeof(Base.literal_pow),Tuple{Base.RefValue{typeof(^)},NTuple{6,Float64},Base.RefValue{Val{2}}}}},Any}}},Zygote.Pullback{Tuple{typeof(sum),NTuple{6,Float64}},Tuple{getfield(Zygote, Symbol("##240#back#147")){getfield(Zygote, Symbol("##145#146")){getfield(Zygote, Symbol("##1705#back#649")){getfield(Zygote, Symbol("##645#647")){NTuple{6,Float64}}},Tuple{NTuple{6,Nothing}}}}}},getfield(Zygote, Symbol("##1021#back#419")){getfield(Zygote, Symbol("##417#418")){Float64}}}},1}}},Tuple{Tuple{Nothing},NTuple{6,Nothing}}}})(::Zygote.FillArray{Float64,1}) at /home/robert/.julia/dev/Zygote/src/lib/grad.jl:41
   [24] #49 at ./none:0 [inlined]
   [25] (::Zygote.Pullback{Tuple{getfield(Main, Symbol("##49#55")),Array{Float64,1},Array{Float64,1},Array{Float64,1},Array{Float64,1},Array{Float64,1},Array{Float64,1}},Any})(::Float64) at /home/robert/.julia/dev/Zygote/src/compiler/interface2.jl:0
   [26] (::getfield(Zygote, Symbol("##145#146")){Zygote.Pullback{Tuple{getfield(Main, Symbol("##49#55")),Array{Float64,1},Array{Float64,1},Array{Float64,1},Array{Float64,1},Array{Float64,1},Array{Float64,1}},Any},Tuple{NTuple{6,Nothing}}})(::Float64) at /home/robert/.julia/dev/Zygote/src/lib/lib.jl:81
   [27] (::getfield(Zygote, Symbol("##240#back#147")){getfield(Zygote, Symbol("##145#146")){Zygote.Pullback{Tuple{getfield(Main, Symbol("##49#55")),Array{Float64,1},Array{Float64,1},Array{Float64,1},Array{Float64,1},Array{Float64,1},Array{Float64,1}},Any},Tuple{NTuple{6,Nothing}}}})(::Float64) at /home/robert/.julia/dev/Zygote/src/lib/grad.jl:41
   [28] #3 at /home/robert/.julia/dev/Zygote/test/gradcheck.jl:25 [inlined]
   [29] (::Zygote.Pullback{Tuple{getfield(Main, Symbol("##3#4")){getfield(Main, Symbol("##49#55"))},Array{Float64,1},Array{Float64,1},Array{Float64,1},Array{Float64,1},Array{Float64,1},Array{Float64,1}},Any})(::Int8) at /home/robert/.julia/dev/Zygote/src/compiler/interface2.jl:0
   [30] (::getfield(Zygote, Symbol("##72#73")){Zygote.Pullback{Tuple{getfield(Main, Symbol("##3#4")){getfield(Main, Symbol("##49#55"))},Array{Float64,1},Array{Float64,1},Array{Float64,1},Array{Float64,1},Array{Float64,1},Array{Float64,1}},Any}})(::Int8) at /home/robert/.julia/dev/Zygote/src/compiler/interface.jl:38
   [31] gradient(::Function, ::Array{Float64,1}, ::Vararg{Any,N} where N) at /home/robert/.julia/dev/Zygote/src/compiler/interface.jl:44
   [32] gradcheck(::Function, ::Array{Float64,1}, ::Vararg{Any,N} where N) at /home/robert/.julia/dev/Zygote/test/gradcheck.jl:21
   [33] gradtest(::Function, ::Array{Float64,1}, ::Vararg{AbstractArray,N} where N) at /home/robert/.julia/dev/Zygote/test/gradcheck.jl:25
   [34] macro expansion at /home/robert/.julia/dev/Zygote/test/gradcheck.jl:75 [inlined]
   [35] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.0/Test/src/Test.jl:1083 [inlined]
   [36] top-level scope at /home/robert/.julia/dev/Zygote/test/gradcheck.jl:74
┌ Warning: implicit `dims=2` argument now has to be passed explicitly to specify that distances between columns should be computed
│   caller = ip:0x0
└ @ Core :-1
Compiler: Error During Test at /home/robert/.julia/dev/Zygote/test/compiler.jl:6
  Test threw exception
  Expression: begin
    res = #= /home/robert/.julia/dev/Zygote/test/compiler.jl:7 =# @inferred(forward(pow, 2, 3))
    true
end
  UndefVarError: pow not defined
  Stacktrace:
   [1] top-level scope at /home/robert/.julia/dev/Zygote/test/compiler.jl:6
   [2] include at ./boot.jl:317 [inlined]
   [3] include_relative(::Module, ::String) at ./loading.jl:1038
   [4] include(::Module, ::String) at ./sysimg.jl:29
   [5] include(::String) at ./client.jl:388
   [6] top-level scope at /home/robert/.julia/dev/Zygote/test/compiler.jl:81
   [7] include at ./boot.jl:317 [inlined]
   [8] include_relative(::Module, ::String) at ./loading.jl:1038
   [9] include(::Module, ::String) at ./sysimg.jl:29
   [10] include(::String) at ./client.jl:388
   [11] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.0/Test/src/Test.jl:1092 [inlined]
   [12] macro expansion at /home/robert/.julia/dev/Zygote/test/runtests.jl:7 [inlined]
   [13] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.0/Test/src/Test.jl:1083 [inlined]
   [14] top-level scope at /home/robert/.julia/dev/Zygote/test/runtests.jl:6
Compiler: Error During Test at /home/robert/.julia/dev/Zygote/test/compiler.jl:6
  Test threw exception
  Expression: begin
    res = #= /home/robert/.julia/dev/Zygote/test/compiler.jl:7 =# @inferred(dpow(2, 3))
    true
end
  UndefVarError: pow not defined
  Stacktrace:
   [1] dpow(::Int64, ::Int64) at /home/robert/.julia/dev/Zygote/test/typed.jl:4
   [2] top-level scope at /home/robert/.julia/dev/Zygote/test/compiler.jl:6
   [3] include at ./boot.jl:317 [inlined]
   [4] include_relative(::Module, ::String) at ./loading.jl:1038
   [5] include(::Module, ::String) at ./sysimg.jl:29
   [6] include(::String) at ./client.jl:388
   [7] top-level scope at /home/robert/.julia/dev/Zygote/test/compiler.jl:81
   [8] include at ./boot.jl:317 [inlined]
   [9] include_relative(::Module, ::String) at ./loading.jl:1038
   [10] include(::Module, ::String) at ./sysimg.jl:29
   [11] include(::String) at ./client.jl:388
   [12] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.0/Test/src/Test.jl:1092 [inlined]
   [13] macro expansion at /home/robert/.julia/dev/Zygote/test/runtests.jl:7 [inlined]
   [14] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.0/Test/src/Test.jl:1083 [inlined]
   [15] top-level scope at /home/robert/.julia/dev/Zygote/test/runtests.jl:6
Compiler: Error During Test at /home/robert/.julia/dev/Zygote/test/compiler.jl:6
  Test threw exception
  Expression: begin
    res = #= /home/robert/.julia/dev/Zygote/test/compiler.jl:7 =# @inferred(forward(cube, 2))
    true
end
  UndefVarError: pow not defined
  Stacktrace:
   [1] cube at /home/robert/.julia/dev/Zygote/test/typed.jl:9 [inlined]
   [2] _forward(::Zygote.Context, ::typeof(cube), ::Int64) at /home/robert/.julia/dev/Zygote/src/compiler/interface2.jl:0
   [3] _forward(::Function, ::Int64) at /home/robert/.julia/dev/Zygote/src/compiler/interface.jl:31
   [4] forward(::Function, ::Int64) at /home/robert/.julia/dev/Zygote/src/compiler/interface.jl:37
   [5] top-level scope at /home/robert/.julia/dev/Zygote/test/compiler.jl:6
   [6] include at ./boot.jl:317 [inlined]
   [7] include_relative(::Module, ::String) at ./loading.jl:1038
   [8] include(::Module, ::String) at ./sysimg.jl:29
   [9] include(::String) at ./client.jl:388
   [10] top-level scope at /home/robert/.julia/dev/Zygote/test/compiler.jl:81
   [11] include at ./boot.jl:317 [inlined]
   [12] include_relative(::Module, ::String) at ./loading.jl:1038
   [13] include(::Module, ::String) at ./sysimg.jl:29
   [14] include(::String) at ./client.jl:388
   [15] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.0/Test/src/Test.jl:1092 [inlined]
   [16] macro expansion at /home/robert/.julia/dev/Zygote/test/runtests.jl:7 [inlined]
   [17] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.0/Test/src/Test.jl:1083 [inlined]
   [18] top-level scope at /home/robert/.julia/dev/Zygote/test/runtests.jl:6
Compiler: Error During Test at /home/robert/.julia/dev/Zygote/test/runtests.jl:14
  Got exception outside of a @test
  LoadError: LoadError: MethodError: no method matching iterate(::Nothing)
  Closest candidates are:
    iterate(!Matched::Core.SimpleVector) at essentials.jl:578
    iterate(!Matched::Core.SimpleVector, !Matched::Any) at essentials.jl:578
    iterate(!Matched::ExponentialBackOff) at error.jl:171
    ...
  Stacktrace:
   [1] indexed_iterate(::Nothing, ::Int64) at ./tuple.jl:63
   [2] top-level scope at /home/robert/.julia/dev/Zygote/test/compiler.jl:10
   [3] include at ./boot.jl:317 [inlined]
   [4] include_relative(::Module, ::String) at ./loading.jl:1038
   [5] include(::Module, ::String) at ./sysimg.jl:29
   [6] include(::String) at ./client.jl:388
   [7] top-level scope at /home/robert/.julia/dev/Zygote/test/compiler.jl:81
   [8] include at ./boot.jl:317 [inlined]
   [9] include_relative(::Module, ::String) at ./loading.jl:1038
   [10] include(::Module, ::String) at ./sysimg.jl:29
   [11] include(::String) at ./client.jl:388
   [12] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.0/Test/src/Test.jl:1092 [inlined]
   [13] macro expansion at /home/robert/.julia/dev/Zygote/test/runtests.jl:7 [inlined]
   [14] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.0/Test/src/Test.jl:1083 [inlined]
   [15] top-level scope at /home/robert/.julia/dev/Zygote/test/runtests.jl:6
   [16] include at ./boot.jl:317 [inlined]
   [17] include_relative(::Module, ::String) at ./loading.jl:1038
   [18] include(::Module, ::String) at ./sysimg.jl:29
   [19] include(::String) at ./client.jl:388
   [20] top-level scope at none:0
   [21] eval(::Module, ::Any) at ./boot.jl:319
   [22] macro expansion at ./logging.jl:317 [inlined]
   [23] exec_options(::Base.JLOptions) at ./client.jl:219
   [24] _start() at ./client.jl:421
  in expression starting at /home/robert/.julia/dev/Zygote/test/typed.jl:11
  in expression starting at /home/robert/.julia/dev/Zygote/test/compiler.jl:80
```

roberthoenig commented:
I can't tell if this is an error with `@adjoint map` or with `usetyped`, and I can't think of a simplified function to verify either option either.

MikeInnes commented:
It may just be that you've hit another unrelated bug in the typed compiler stuff. I'll take a look and we can merge this once I've fixed it.

MikeInnes commented:
Digging through there's a couple of issues: first is that this definition of `map` is a little wide, and catches things like `map(broadcastable, ())`. It should handle that but doesn't actually need to since there's no mutation involved. For now, I've just restricted to `AbstractArray`, which is probably too narrow but we can widen as needed.

Secondly we can't compile `getindex(::Int,::Int)` in typed mode because of the `unreachable` block. We'll have to fix that properly at some point but for now I've just given it a manual adjoint.

MikeInnes commented:
That got it; thanks a lot for the patch @roberthoenig!
